frame/staking/src/lib.rs

```diff
@@ -570,36 +570,46 @@ impl<T: Config> StakingLedger<T> {
 		// (assuming 28 is the bonding duration) onwards should be slashed.
 		let slashable_chunks_start = slash_era + T::BondingDuration::get();
 
-		// `Some(ratio)` if this is proportional, with `ratio`, `None` otherwise. In both cases, we
-		// slash first the active chunk, and then `slash_chunks_priority`.
-		let (maybe_proportional, slash_chunks_priority) = {
-			if let Some(first_slashable_index) =
-				self.unlocking.iter().position(|c| c.era >= slashable_chunks_start)
-			{
-				// If there exists a chunk who's after the first_slashable_start, then this is a
-				// proportional slash, because we want to slash active and these chunks
-				// proportionally.
-
-				// The indices of the first chunk after the slash up through the most recent chunk.
-				// (The most recent chunk is at greatest from this era)
-				let affected_indices = first_slashable_index..self.unlocking.len();
-				let unbonding_affected_balance =
-					affected_indices.clone().fold(BalanceOf::<T>::zero(), |sum, i| {
-						if let Some(chunk) = self.unlocking.get(i).defensive() {
-							sum.saturating_add(chunk.value)
-						} else {
-							sum
-						}
-					});
-				let affected_balance = self.active.saturating_add(unbonding_affected_balance);
-				let ratio = Perquintill::from_rational(slash_amount, affected_balance);
-				(
-					Some(ratio),
-					affected_indices.chain((0..first_slashable_index).rev()).collect::<Vec<_>>(),
-				)
-			} else {
-				// We just slash from the last chunk to the most recent one, if need be.
-				(None, (0..self.unlocking.len()).rev().collect::<Vec<_>>())
+		// if `remaining_ratio` is not zero this is a proportional slash, each chunk and
+		// `self.active` will be slashed proportionally based on the `remaining_ratio`, slashed to
+		// zero otherwise. In both cases, we slash first `self.active`, and then
+		// `slash_chunks_priority`.
+		let (remaining_ratio, slash_chunks_priority) = {
+			let mut affected_balance = BalanceOf::<T>::zero();
```

Comment:
I don't quite agree with the changes here, like `let mut first_chunk = None;` and assigning to it in the loop. This is actually not Rust-idiomatic at all, and the previous `if let Some(_) = iterator {} else {}` is more preferred.

Reply:
I thought using a single loop to get `first_index` and `affected_balance` was more efficient (less iteration) and cleaner (less `if else` branch), but indeed hand code the functionality that the `Iterator` API already provided is not Rust-idiomatic.